### PR TITLE
Assignment3

### PR DESCRIPTION
**Pull request by alexandracalota**
Should have created a pull request, but I only committed the files to a new branch.

### CODE REVIEW

File: Dictionary.java [hunk 1]
```diff
@@ -0,0 +1,49 @@
...
+
+	public void setWords(LinkedList<String> words) {
+		this.words = words;
+	}
+
+	public LinkedList<String> getPrefixes() {
+		return prefixes;
+	}
+
+	public void setPrefixes(LinkedList<String> prefixes) {
```
Comment (kkasiacodeu) [inline Dictionary.java]: You don't want this to be settable at all. This should not be mutable from outside of this class at all.

File: Dictionary.java [hunk 2]
```diff
@@ -0,0 +1,49 @@
...
+
+	public boolean isPrefix(String prefix) {
+		return prefixes.contains(prefix);
+	}
+
+	public LinkedList<String> getWords() {
+		return words;
+	}
+
+	public void setWords(LinkedList<String> words) {
```
Comment (kkasiacodeu) [inline Dictionary.java]: You don't want this to be settable externally. If this is set some time after the creation of the class instance, the prefixes will not be computed for it.

File: Dictionary.java [hunk 3]
```diff
@@ -0,0 +1,49 @@
+import java.util.LinkedList;
+
+public class Dictionary {
+	private LinkedList<String> words;
+	private LinkedList<String> prefixes;
+
+	public Dictionary(LinkedList<String> words) {
+		this.words = words;
+		this.prefixes = new LinkedList<String>();
```
Comment (kkasiacodeu) [inline Dictionary.java]: I would suggest extracting the computing part outside of the constructor. It is generally bad practice to have any work done in the constructor. I suggest you arrange it in the following way: the constructor only initializes the required fields, then we provide a method that sets "words" list and computes "prefixes" list.

File: SearchWord.java [hunk 4]
```diff
@@ -0,0 +1,59 @@
+import java.util.LinkedList;
+
+public class SearchWord {
+
+	public int[][] visited;
```
Comment (kkasiacodeu) [inline SearchWord.java]: this can be a matrix of booleans instead of ints.

File: SearchWord.java [hunk 5]
```diff
@@ -0,0 +1,59 @@
...
+			for (int j = 0; j < columns; j++) {
+				applyBFS(rows, columns, i, j, characters, dict, validWords, "");
+			}
+		}
+
+		return validWords;
+	}
+
+	public void applyBFS(int rows, int columns, int posX, int posY, char[][] characters, Dictionary dict,
+			LinkedList<String> validWords, String word) {
```
Comment (kkasiacodeu) [inline SearchWord.java]: Why is validWords a LinkedList? You perform two operations on it: contains(word) and add(word). There are better data structures to handle these. What is the complexity for those operations for the linked list?

File: Dictionary.java [hunk 6]
```diff
@@ -0,0 +1,49 @@
+import java.util.LinkedList;
+
+public class Dictionary {
+	private LinkedList<String> words;
```
Comment (kkasiacodeu) [inline Dictionary.java]: Could this be a different data structure than LinkedList (better for the complexity of performed operations)?

File: Dictionary.java [hunk 7]
```diff
@@ -0,0 +1,49 @@
+import java.util.LinkedList;
+
+public class Dictionary {
+	private LinkedList<String> words;
+	private LinkedList<String> prefixes;
```
Comment (kkasiacodeu) [inline Dictionary.java]: Could this be a different data structure than LinkedList (better for the complexity of performed operations)?

File: SearchWord.java [hunk 8]
```diff
@@ -0,0 +1,59 @@
...
+		for (int i = 0; i < rows; i++) {
+			for (int j = 0; j < columns; j++) {
+				applyBFS(rows, columns, i, j, characters, dict, validWords, "");
+			}
+		}
+
+		return validWords;
+	}
+
+	public void applyBFS(int rows, int columns, int posX, int posY, char[][] characters, Dictionary dict,
```
Comment (kkasiacodeu) [inline SearchWord.java]: This is an implementation of DFS not BFS :)

**Review by kkasiacodeu — CHANGES_REQUESTED** [review]
The algorithm is good, well done! :)

Just some improvements to the code quality and some data structures used.

But generally great job 👍

File: Dictionary.java (hunk 6 above)
Comment (alexandracalota) [reply Dictionary.java]: Yes, HashSet is faster thank LinkedList. I have changed the data structures to that.

File: Dictionary.java (hunk 3 above)
Comment (alexandracalota) [reply Dictionary.java]: I see, thank you for your advice. I've changed the constructor, please let me know if it's alright now.

File: SearchWord.java (hunk 5 above)
Comment (alexandracalota) [reply SearchWord.java]: I would say contains(word) has O(n) complexity and add(word) has O(1). A HashSet is better, as both contains and add have O(1) complexity.

**Pull request merged**